### PR TITLE
Rewrite multiplex tests to use the memory transport

### PR DESCRIPTION
**Pull request by tomaka**

### CODE REVIEW

File: core/tests/multiplex.rs [hunk 1]
```diff
@@ -74,19 +73,14 @@ fn client_to_server_outbound() {
...
-
-        let (listener, addr) = transport
-            .listen_on("/ip4/127.0.0.1/tcp/0".parse().unwrap())
-            .unwrap_or_else(|_| panic!());
-        tx.send(addr).unwrap();
-
-        let future = listener
+            .into_connection_reuse()
+            .listen_on("/memory".parse().unwrap())
+            .unwrap_or_else(|_| panic!()).0
```
Comment (gnunicorn) [inline core/tests/multiplex.rs]: why not just `unwrap()`?

File: core/tests/multiplex.rs [hunk 2]
```diff
@@ -107,11 +101,10 @@ fn client_to_server_outbound() {
...
 
-    let transport = TcpConfig::new().with_upgrade(multiplex::MplexConfig::new());
-
-    let future = transport
-        .dial(rx.recv().unwrap())
-        .unwrap()
+    let future = tx
+        .with_upgrade(multiplex::MplexConfig::new())
+        .dial("/memory".parse().unwrap())
+        .unwrap_or_else(|_| panic!())
```
Comment (gnunicorn) [inline core/tests/multiplex.rs]: same

File: core/tests/multiplex.rs [hunk 3]
```diff
@@ -126,19 +119,14 @@ fn connection_reused_for_dialing() {
...
-
-        let (listener, addr) = transport
-            .listen_on("/ip4/127.0.0.1/tcp/0".parse().unwrap())
-            .unwrap_or_else(|_| panic!());
-        tx.send(addr).unwrap();
-
-        let future = listener
+            .into_connection_reuse()
+            .listen_on("/memory".parse().unwrap())
+            .unwrap_or_else(|_| panic!()).0
```
Comment (gnunicorn) [inline core/tests/multiplex.rs]: same as above.

File: core/tests/multiplex.rs [hunk 4]
```diff
@@ -170,22 +158,20 @@ fn connection_reused_for_dialing() {
...
+        .dial("/memory".parse().unwrap())
         .unwrap_or_else(|_| panic!())
         .map(|server| Framed::<_, BytesMut>::new(server.0))
         .and_then(|server| server.send("hello world".into()))
         .and_then(|first_connec| {
             transport
                 .clone()
-                .dial(listen_addr.clone())
+                .dial("/memory".parse().unwrap())
                 .unwrap_or_else(|_| panic!())
```
Comment (gnunicorn) [inline core/tests/multiplex.rs]: I guess you took it from here? Still unwrap already panics, why `panic!` in the `or_else` - that doesn't make any sense to me...?

File: core/tests/multiplex.rs (hunk 1 above)
Comment (tomaka) [reply core/tests/multiplex.rs]: The return value doesn't implement `Debug`. See #367

File: core/tests/multiplex.rs (hunk 4 above)
Comment (tomaka) [reply core/tests/multiplex.rs]: I'm not sure if that's what you mean, but the `unwrap()` applies to parsing `/memory` into a `Multiaddr`.

**Pull request merged**